File: agents/python/crypto/aes.py
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

NONCE_SIZE = 12
KEY_SIZE = 32
TAG_SIZE = 16


class AESCipher:
    """AES-256-GCM authenticated encryption."""

    def __init__(self, key: bytes):
        if len(key) != KEY_SIZE:
            raise ValueError(f"Key must be {KEY_SIZE} bytes, got {len(key)}")
        self._gcm = AESGCM(key)

    def encrypt(self, plaintext: bytes) -> bytes:
        """Returns nonce(12) + ciphertext + tag(16)."""
        nonce = os.urandom(NONCE_SIZE)
        ct = self._gcm.encrypt(nonce, plaintext, None)
        return nonce + ct

    def decrypt(self, data: bytes) -> bytes:
        """Expects nonce(12) + ciphertext + tag(16)."""
        if len(data) < NONCE_SIZE + TAG_SIZE:
            raise ValueError("Ciphertext too short")
        nonce = data[:NONCE_SIZE]
        ct = data[NONCE_SIZE:]
        return self._gcm.decrypt(nonce, ct, None)

    @staticmethod
    def generate_key() -> bytes:
        return os.urandom(KEY_SIZE)

File: agents/go/crypto/aes.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
	"io"
)

const (
	KeySize   = 32 // AES-256
	NonceSize = 12 // GCM standard nonce
)

// GenerateKey returns a cryptographically random 32-byte AES-256 key.
func GenerateKey() ([]byte, error) {
	key := make([]byte, KeySize)
	if _, err := io.ReadFull(rand.Reader, key); err != nil {
		return nil, fmt.Errorf("generate key: %w", err)
	}
	return key, nil
}

// Encrypt performs AES-256-GCM encryption.
// Returns: nonce(12) || ciphertext || tag (GCM appends tag to ciphertext).
func Encrypt(key, plaintext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("new gcm: %w", err)
	}

	nonce := make([]byte, NonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, fmt.Errorf("generate nonce: %w", err)
	}

	// Seal appends ciphertext+tag to nonce prefix
	ciphertext := gcm.Seal(nonce, nonce, plaintext, nil)
	return ciphertext, nil
}

// Decrypt performs AES-256-GCM decryption.
// Expects input format: nonce(12) || ciphertext || tag.
func Decrypt(key, data []byte) ([]byte, error) {
	if len(data) < NonceSize {
		return nil, fmt.Errorf("ciphertext too short")
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("new gcm: %w", err)
	}

	nonce := data[:NonceSize]
	ciphertext := data[NonceSize:]

	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}
	return plaintext, nil
}
